src/cargo/core/compiler/fingerprint.rs

```diff
@@ -350,13 +350,23 @@ use super::{BuildContext, Context, FileFlavor, Unit};
 // While source files can't currently be > 4Gb, bin files could be.
 pub type FileSize = u64;
 
-//TODO: implement hash yourself
-#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize, Hash)]
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
 pub struct FileHash {
     pub kind: FileHashAlgorithm,
     pub hash: String,
 }
 
+impl hash::Hash for FileHash {
+    fn hash<H>(&self, hasher: &mut H)
+    where
+        H: hash::Hasher,
+    {
+        for ch in self.hash.as_bytes() {
```

Comment:
Why did you omit `kind`?

Reply:
I thought I should implement hash myself as there was no point hashing a hash, but even if I implement hash all I can do is add the hash to the data to be hashed so I think I should just derive Hash here - I don't think there's an optimisation to be had.